https://www.kaggle.com/littlemountainman/tuned-point-to-uncertainty

In [1]:
!pip install pandas --upgrade --no-deps
import pandas as pd
pd.__version__

Requirement already up-to-date: pandas in c:\programdata\anaconda3\lib\site-packages (1.0.4)


'1.0.4'

In [1]:
import pandas as pd, numpy as np
from matplotlib import pyplot as plt

import scipy.stats  as stats

In [2]:
pd.options.display.max_columns = 50

In [4]:
best = pd.read_csv("submission_kaggle_v2.csv")
best.head()

id        F1        F2        F3        F4  \
0  HOBBIES_1_001_CA_1_validation  0.761828  0.707905  0.719281  0.751882   
1  HOBBIES_1_002_CA_1_validation  0.193685  0.186511  0.181866  0.179773   
2  HOBBIES_1_003_CA_1_validation  0.449895  0.420316  0.416436  0.414502   
3  HOBBIES_1_004_CA_1_validation  1.636948  1.272583  1.326518  1.511982   
4  HOBBIES_1_005_CA_1_validation  0.951627  0.850306  0.855925  0.933599   

         F5        F6        F7        F8        F9       F10       F11  \
0  0.927137  1.017785  1.144739  0.849844  0.865100  0.770670  0.743174   
1  0.221516  0.264652  0.294685  0.237230  0.224753  0.188160  0.228585   
2  0.578287  0.742792  0.758611  0.492596  0.501036  0.446619  0.471248   
3  1.909681  2.879874  3.175579  1.766516  1.495614  1.423713  1.570031   
4  1.063057  1.470444  1.511544  0.977362  0.978745  0.983190  0.909601   

        F12       F13       F14       F15       F16       F17       F18  \
0  0.901572  1.077072  0.846744  0.857600  0.826266  0.756897  0.741843   
1  0.224789  0.228196  0.223050  0.227482  0.201790  0.183923  0.189119   
2  0.536249  0.682463  0.551578  0.447078  0.422443  0.429068  0.465357   
3  1.777553  2.942041  2.236004  1.684949  1.480747  1.343602  1.407639   
4  1.011580  1.458266  1.064466  0.994964  1.059548  1.005258  1.074282   

        F19       F20       F21       F22       F23       F24       F25  \
0  0.778617  0.962237  0.983039  0.829216  0.773040  0.738191  0.788460   
1  0.209571  0.227181  0.229576  0.198796  0.181461  0.188745  0.182155   
2  0.555078  0.714344  0.718341  0.454664  0.425591  0.454029  0.485734   
3  1.879871  2.726490  3.310782  1.729650  1.479737  1.477537  1.453358   
4  1.049718  1.629175  1.630603  1.094082  0.881098  0.982165  0.949721   

        F26       F27       F28  
0  0.894650  1.055066  0.963816  
1  0.196674  0.237201  0.252933  
2  0.621030  0.738850  0.698694  
3  2.096524  3.207611  3.472014  
4  1.125167  1.604738  1.547678

In [5]:
sales = pd.read_csv("sales_train_validation.csv")
sales.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  d_5  d_6  d_7  d_8  d_9  d_10  d_11  d_12  \
0       CA    0    0    0    0    0    0    0    0    0     0     0     0   
1       CA    0    0    0    0    0    0    0    0    0     0     0     0   
2       CA    0    0    0    0    0    0    0    0    0     0     0     0   
3       CA    0    0    0    0    0    0    0    0    0     0     0     0   
4       CA    0    0    0    0    0    0    0    0    0     0     0     0   

   d_13  d_14  d_15  d_16  d_17  d_18  d_19  ...  d_1889  d_1890  d_1891  \
0     0     0     0     0     0     0     0  ...       0       0       0   
1     0     0     0     0     0     0     0  ...       0       0       0   
2     0     0     0     0     0     0     0  ...       0       0       0   
3     0     0     0     0     0     0     0  ...       0       3       1   
4     0     0     0     0     0     0     0  ...       4       0       1   

   d_1892  d_1893  d_1894  d_1895  d_1896  d_1897  d_1898  d_1899  d_1900  \
0       1       0       4       2       3       0       1       2       0   
1       0       0       0       0       0       0       0       0       0   
2       1       0       0       0       1       0       0       0       0   
3       2       1       3       1       0       2       5       4       2   
4       4       0       1       0       1       0       1       1       2   

   d_1901  d_1902  d_1903  d_1904  d_1905  d_1906  d_1907  d_1908  d_1909  \
0       0       0       1       1       3       0       1       1       1   
1       0       0       0       0       0       0       0       0       1   
2       0       1       2       2       1       2       1       1       1   
3       0       3       0       1       0       5       4       1       0   
4       0       1       1       2       1       1       0       1       1   

   d_1910  d_1911  d_1912  d_1913  
0       3       0       1       1  
1       0       0       0       0  
2       0       1       1       1  
3       1       3       7       2  
4       2       2       2       4  

[5 rows x 1919 columns]

In [6]:
sub = best.merge(sales[["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"]], on = "id")
sub["_all_"] = "Total"
sub.shape

(30490, 35)

In [7]:
sub.head()

id        F1        F2        F3        F4  \
0  HOBBIES_1_001_CA_1_validation  0.761828  0.707905  0.719281  0.751882   
1  HOBBIES_1_002_CA_1_validation  0.193685  0.186511  0.181866  0.179773   
2  HOBBIES_1_003_CA_1_validation  0.449895  0.420316  0.416436  0.414502   
3  HOBBIES_1_004_CA_1_validation  1.636948  1.272583  1.326518  1.511982   
4  HOBBIES_1_005_CA_1_validation  0.951627  0.850306  0.855925  0.933599   

         F5        F6        F7        F8        F9       F10       F11  \
0  0.927137  1.017785  1.144739  0.849844  0.865100  0.770670  0.743174   
1  0.221516  0.264652  0.294685  0.237230  0.224753  0.188160  0.228585   
2  0.578287  0.742792  0.758611  0.492596  0.501036  0.446619  0.471248   
3  1.909681  2.879874  3.175579  1.766516  1.495614  1.423713  1.570031   
4  1.063057  1.470444  1.511544  0.977362  0.978745  0.983190  0.909601   

        F12       F13       F14       F15       F16       F17       F18  \
0  0.901572  1.077072  0.846744  0.857600  0.826266  0.756897  0.741843   
1  0.224789  0.228196  0.223050  0.227482  0.201790  0.183923  0.189119   
2  0.536249  0.682463  0.551578  0.447078  0.422443  0.429068  0.465357   
3  1.777553  2.942041  2.236004  1.684949  1.480747  1.343602  1.407639   
4  1.011580  1.458266  1.064466  0.994964  1.059548  1.005258  1.074282   

        F19       F20       F21       F22       F23       F24       F25  \
0  0.778617  0.962237  0.983039  0.829216  0.773040  0.738191  0.788460   
1  0.209571  0.227181  0.229576  0.198796  0.181461  0.188745  0.182155   
2  0.555078  0.714344  0.718341  0.454664  0.425591  0.454029  0.485734   
3  1.879871  2.726490  3.310782  1.729650  1.479737  1.477537  1.453358   
4  1.049718  1.629175  1.630603  1.094082  0.881098  0.982165  0.949721   

        F26       F27       F28        item_id    dept_id   cat_id store_id  \
0  0.894650  1.055066  0.963816  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  0.196674  0.237201  0.252933  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  0.621030  0.738850  0.698694  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  2.096524  3.207611  3.472014  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  1.125167  1.604738  1.547678  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  _all_  
0       CA  Total  
1       CA  Total  
2       CA  Total  
3       CA  Total  
4       CA  Total

Different ratios for different aggregation levels
The higher the aggregation level, the more confident we are in the point prediction --> lower coef, relatively smaller range of quantiles

In [8]:
qs = np.array([0.005,0.025,0.165,0.25, 0.5, 0.75, 0.835, 0.975, 0.995])
qs.shape


def get_ratios(coef=0.15):
    qs2 = np.log(qs/(1-qs))*coef
    ratios = stats.norm.cdf(qs2)
    ratios /= ratios[4]
    ratios = pd.Series(ratios, index=qs)
    return ratios.round(3)

#coef between 0.05 and 0.24 is used, probably suboptimal values for now

level_coef_dict = {"id": get_ratios(coef=0.17), "item_id": get_ratios(coef=0.10),
                   "dept_id": get_ratios(coef=0.04), "cat_id": get_ratios(coef=0.04),
                   "store_id": get_ratios(coef=0.04), "state_id": get_ratios(coef=0.04), "_all_": get_ratios(coef=0.04),
                   ("state_id", "item_id"): get_ratios(coef=0.12),  ("state_id", "dept_id"): get_ratios(coef=0.08),
                    ("store_id","dept_id") : get_ratios(coef=0.05), ("state_id", "cat_id"): get_ratios(coef=0.06),
                    ("store_id","cat_id"): get_ratios(coef=0.1)
                  }

In [9]:
level_coef_dict["id"]

0.005    0.368
0.025    0.533
0.165    0.783
0.250    0.852
0.500    1.000
0.750    1.148
0.835    1.217
0.975    1.467
0.995    1.632
dtype: float64

In [10]:
level_coef_dict["cat_id"]

0.005    0.832
0.025    0.883
0.165    0.948
0.250    0.965
0.500    1.000
0.750    1.035
0.835    1.052
0.975    1.117
0.995    1.168
dtype: float64

For the the lowest level (30490 series), the smallest and biggest quantiles are 20% and 180% of the point prediction. For categories (3 series), the model will be way more confident: the smallest quantile will be 71%, the biggest will be 129% of the point prediction.

In [11]:
def quantile_coefs(q, level):
    ratios = level_coef_dict[level]        
    return ratios.loc[q].values

def get_group_preds(pred, level):
    df = pred.groupby(level)[cols].sum()
    q = np.repeat(qs, len(df))
    df = pd.concat([df]*9, axis=0, sort=False)
    df.reset_index(inplace = True)
    df[cols] *= quantile_coefs(q, level)[:, None]
    if level != "id":
        df["id"] = [f"{lev}_X_{q:.3f}_validation" for lev, q in zip(df[level].values, q)]
    else:
        df["id"] = [f"{lev.replace('_validation', '')}_{q:.3f}_validation" for lev, q in zip(df[level].values, q)]
    df = df[["id"]+list(cols)]
    return df

def get_couple_group_preds(pred, level1, level2):
    df = pred.groupby([level1, level2])[cols].sum()
    q = np.repeat(qs, len(df))
    df = pd.concat([df]*9, axis=0, sort=False)
    df.reset_index(inplace = True)
    df[cols] *= quantile_coefs(q, (level1, level2))[:, None]
    df["id"] = [f"{lev1}_{lev2}_{q:.3f}_validation" for lev1,lev2, q in 
                zip(df[level1].values,df[level2].values, q)]
    df = df[["id"]+list(cols)]
    return df

levels = ["id", "item_id", "dept_id", "cat_id", "store_id", "state_id", "_all_"]
couples = [("state_id", "item_id"),  ("state_id", "dept_id"),("store_id","dept_id"),
                            ("state_id", "cat_id"),("store_id","cat_id")]
cols = [f"F{i}" for i in range(1, 29)]

df = []
for level in levels :
    df.append(get_group_preds(sub, level))
for level1,level2 in couples:
    df.append(get_couple_group_preds(sub, level1, level2))
df = pd.concat(df, axis=0, sort=False)
df.reset_index(drop=True, inplace=True)
df = pd.concat([df,df] , axis=0, sort=False)
df.reset_index(drop=True, inplace=True)
df.loc[df.index >= len(df.index)//2, "id"] = df.loc[df.index >= len(df.index)//2, "id"].str.replace(
                                    "_validation$", "_evaluation")

df.shape

df.head()

df.to_csv("submission_PD_v2.csv", index = False)